### ISSUE TITLE
Exception if keys are UUID

### ISSUE THREAD

**Issue by sshishov** [claytondaley/drf-keyed-list#6]
I noticed this problem during applying the solution: if the key is UUID then the response dict is populated as UUID values which are not allowed in `json.dumps` function.
Python3 produces this exceptions:
```python
.../3.7/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
TypeError: keys must be str, int, float, bool or None, not UUID
```
Python2 produces this:
```python
.../2.7/lib/python2.7/json/encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
TypeError: keys must be a string
```

I do not know what should we do here. Either we should convert all not supported values to `str` or maybe leave this on to being handled by custom JsonEncoder?!

**Comment by claytondaley** [comment]
JSON keys must be strings.  Go ahead and wrap the key generator in `str()`.

**Comment by sshishov** [comment]
To tell in truth, python3 supports JSON keys as `str`, `int`, `float`, `bool` or `None` as you can see from error message. Therefore it will be not wise to wrap in `str` for all keys IMHO

I wrote something like this instead:
```python
def to_representation(self, data):
    response = super().to_representation(data)
    data = {}
    for value in response:
        key = value.pop(self._keyed_field)
        if isinstance(key, uuid.UUID):
            key = str(key)
        data[key] = value
    return data
```
And we can extend it in future with additional `isinstance` if needed

**Comment by claytondaley** [comment]
`dict` keys may be other things, but according to the [JSON RFC](https://tools.ietf.org/html/rfc7159.html#section-4):

> zero or more name/value pairs (or members).  A name is a string.

Edit:  I'm surprised that the JSON Encoder accepts other types as keys, but I agree that it must given the error text.  I'd be a little worried about using numeric keys as I'm not convinced it would faithfully reproduce types on the other end.

**Comment by claytondaley** [comment]
... though I guess this serializer could be used with other encoders so maybe we need to preserve the object and fix it in the JSON Encoder itself.

**Comment by claytondaley** [comment]
OK.  Did some more thinking.  

The "right" way to fix this is to use a custom serializer field (e.g. `UUIDStringField`).  This field should have `to_*` methods that convert the UUID to (and from) a string.  Use that field on your serializer (or register it as the default serializer for UUID fields).  That will ensure the serializer always returns a string so the JSON encoder has valid keys (FWIW I don't think a UUID is a valid JSON value either).

That's your best option because:

 - You don't want to mess with the JSON Encoder.  It needs to be be fully reversible and shouldn't have to decide whether a string is a UUID (or not) for casting reasons.
 - This package needs to work with other kinds of encoders so it shouldn't cast anything.  The most it should do is to restrict the keys to valid python `dict` keys (i.e. hashable types).
    - As a thought experiment, a `PickleEncoder` might not mind that the key is a `UUID` and might have no problems encoding/decoding the UUID.  Thus we need to preserve the object type.

**Comment by claytondaley** [comment]
BTW I pushed a 0.0.2 with your fixes to pypi.

**Comment by sshishov** [comment]
You are right, I will try to use proper Field for ID field and see if it works

**Comment by sshishov** [comment]
Confirming that it is working. No need additional changes, but better to write the example how to handle such situations because it can happen with any not str field used as a key.

In my case I used ModelSerializer and primary key field, therefore it created behind the scenes based on provided `fields` value:
```python
field = PrimaryKeyRelatedField(queryset=MyModel.objects.all())
```

This field could not properly handle UUID value. Instead I specified the field the following way:
```python
field = serializers.UUIDField(source='field_id')
```
After this everything working without exceptions. You were right. Great that you thought about it.
